### PR TITLE
Custom structures

### PR DESCRIPTION
**Pull request by bdistin**
**Please describe the changes this PR makes and why it should be merged:**
This allows any structure created by a DataStore to be properly extended.

A simple example: 
```js
const { Client, Message } = require('discord.js');

class ExtendedMessage extends Message {

	_patch(data) {
		super._patch(data);
		this.isCommand = this.content.startsWith('%');
	}

}

const client = new Client({ customStructures: { Message: ExtendedMessage } });
```

results: img

More advanced use-case: Commando could simply extend Message and Guild, rather than `ye old lazy dev getters` and applyToClass functions.

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by devsnek** [comment]
your example should use a getter, running string ops on every message is kinda useless

**Comment by appellation** [comment]
First of all, I disagree with not actually modifying the built-in structures.  While pretty much every structure in Discord.js is created through a data store, it feels hack-ish to apply a completely different class in a specific (albeit common) situation; if you're extending a class, it would make sense to extend the class for the entire application in order to mitigate bugs and ensure consistency.  Second, it's not super hard to apply custom properties and methods to JS objects: this functionality already exists within the language, and I believe it is better to be explicit about modifying a built-in class.  Third, built-in structures should only be modified by people who are confident in their understanding of how the underlying library works, and making this easier has the potential to confuse people who do this improperly.

In summary:

* this introduces inconsistencies
* this functionality already exists
* this functionality should not be made easier

**Comment by kyranet** [comment]
@devsnek You are going to process that in your command handler anyways, otherwise how would you recognize if a command is prefixed or not? Also that's an example, another example would be extending the Guild structure to add a property of 'settings', instead of the classic getter that, besides of being incredibly slow (both the getter and the Map#get (for versions before Node.js 9.0.0)), is inefficient.

With this PR, we can easily change the holds rather than doing Object.defineProperty on these stores, but it may lead to several issues when creating new ones. What's the point on making it harder, @appellation. This change allows frameworks to unleash their entire power and boost the performance, if a user who has no idea of how to do this,  tries to (unsuccessfully) replace a structure, it's their fault, there are other things that are optional in discord.js, and however, it's not for good (for example, you can change from secuential to burst), and even, change the http options and such. In short:

- This is a client option, they're not meant to be used with no knowledge of what every property do.
- This can boost performance in many command handlers or frameworks.
- This can also save many instructions. If you have a getter 'settings' attached to Guild and you call it 4 times in every command, then you're processing the whole getter 4 times, which leads to performance issues.

**Comment by MrJacz** [comment]
👌

**Comment by kyranet** [comment]
Frameworks like Klasa or Commando can make a Klasa/CommandGuild class and replace it with this system, however, both frameworks should export their respective classes so we can extend it for our needs (like the frameworks would do with discord.js), then, the developer must have the possibility to replace the class in their respective client options. I.e. if you pass `customStructures: { Guild: ExtendedDevGuild }`, being ExtendedDevGuild a class that extends KlasaGuild (which also extends Guild), the framework must use ExtendedDevGuild rather than KlasaGuild. Same for Commando and any other framework that could benefit from this.

This is, however, not a breaking change, ExtendedDevGuild and Klasa/CommandoGuild are still instances of Guild as they extend it, so SomeLibGuild shouldn't have issues with it. The method appellation suggested that already exists can still be applied.

**Comment by 1Computer1** [comment]
I agree with both appel and Crawl, `Object.defineProperty` exists and other libs should not have to account for and depend on other libs to work properly.

The only two advantages I can think of for this method is that you can have custom behavior in the constructor and be able to do `super.<member>`. However, both of those should be completely uneccessary.

For the constructor, I can think of modifying existing properties or adding more properties. The first should be completely out of the question. You can make a case for the second, but that can be done with getters just fine (if you are that ornery about perf, just make a cached getter).

Being able to do `super.<member>` means that you're overriding a method, which should not be done, in case some other lib or even D.JS itself requires the normal behavior of the method.

In my opinion, this PR is a bad solution to a non-issue.

**Comment by kyranet** [comment]
Problem with `Object.defineProperty` is that you need to apply it for each DataStore instance, but all classes created before this function call will use the old structure rather than the extended one, nonetheless future objects will use the new/extended structure, and it leads to inconsistency, as well as performance issues at startup since we need to iterate over all DataStore instances for something, it's even worse when it comes to extending Message, since we'd need to iterate over each `TextChannel#messages`.

*if you are that ornery about perf, just make a cached getter* You are never doing `this.guild = client.guilds.get(data.guild_id);` for every message as it's useless since `this.channel.guild` is a property (not a getter) whose value is a reference of an instance of Guild. Therefore it takes almost nothing to access to it. About the cached getter, I'm personally using a Map instance to store guild settings, however, Map#get performance was nearly 29M ops/sec before node.js 9.x and over 510M ops/sec after it. However, it's much slower than having it as a property instead of a getter. (Isn't extending instances of classes by assigning them a variable considered a bad practise?)

Being able to do `super.<member>` is dangerous and I agree about that, but as I said, blindly extending structures and resulting on corrupting the functionality of the library is the developer's fault. You can still do

```js
method(...args) {
    super(...args);
    // Other logic, including console.log for debugging purposes
    // and issue troubleshoot
}
```

I know this PR does not fix any issue, but it'd be very nice to have it, and let the developers who know d.js internals be able to take advantage of this to boost their bots.

**Comment by 1Computer1** [comment]
When we said `Object.defineProperty` we meant on `Class.prototype` not on the instances.

**Comment by appellation** [comment]
**Simplicity:** Anyone who knows discord.js internals should also know how to extend a built-in class.

**Performance:** I believe that performance concerns are a non-issue; while I understand that getter benchmarks show a definite decrease from properties, I can't imagine anyone running millions of ops on their custom channel getter (is it that hard to cache it yourself if you're concerned?) to notice a difference.

**Extensibility:** I thought the point of this PR was to avoid the method that I described.  Making it the only possible option for multiple external extensions (e.g. if, for some reason, I wanted to extend both Commando's guild and Akairo's guild) seems like an even worse idea.  If you're going to commit to something, you shouldn't have to fallback to a different solution to make it work.

**Consistency:** If you modify the prototype immediately after requiring Discord.js, I see no reason why any data structures should have been created.  There should be absolutely no reason to iterate over `TextChannel#messages` for this purpose, thus making performance even more of a non-issue.

---

Taking away the sugary `extends` keyword, let's look at what we've got.  Rather than directly modifying the prototype of the class we want to enhance, we're essentially copying the built-in class to a new class, modifying it, and then monkey patching it into the library so this completely different class gets used most of the time.  To me this feels like a hack to use ES6 features in a place where they should not be used.

**Comment by bdistin** [comment]
Only those who create a new DiscordClient really have a say in what the customStructures are. Generally speaking, i would have klasa respect that if you pass your own custom classes in the d.js options portion of the klasaClientOptions. If it didn't work, they will find out quickly, just as anyone who would do it to d.js directly. I assume commando would do the same, although I also suppose you wouldn't have to respect that option. It's really down to the framework. But to my first point, if someone was using Commando and Discord.js-Music, d.js-music wouldn't even have a say because it doesn't create a d.js client.

But, at the framework level, a dev can make their own class, the framework accepts that they may screw up but that's on them and pass it on to d.js. A dev knowing what they are doing would extend the frameworks extended class, and pass the extended extended class in.

**Comment by bdistin** [comment]
Closing in favor of @Gawdl3y 's custom-structures branch which takes this a step further.
